### PR TITLE
CHAT-984: Apply override mecanism

### PR DESCRIPTION
**Pull request by alaameddeb**

### CODE REVIEW

File: application/src/main/webapp/vue-app/extension.js [hunk 1]
```diff
@@ -539,10 +537,19 @@ export const DEFAULT_MESSAGE_ACTIONS = [
...
-}
\ No newline at end of file
+  return defaultExtensions;
+}
+
+export const extraMessageTypes = getExtenstionByType('message-types');
+export const extraMessageNotifs = getExtenstionByType('message-notifs');
+export const composerApplications = mergeExtensions(DEFAULT_COMPOSER_APPS, getExtenstionByType('composer-applications'));
+export const messageActions = mergeExtensions(DEFAULT_MESSAGE_ACTIONS, getExtenstionByType('message-actions'));
+export const roomActions = mergeExtensions(DEFAULT_ROOM_ACTIONS, getExtenstionByType('room-actions'));
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/vue-app/extension.js]: I would set the extension type in singular ('message-type', 'message-notif', ...). It is the name of the type, sounds better in singular IMO.

File: application/src/main/webapp/vue-app/extension.js [hunk 2]
```diff
@@ -539,10 +537,19 @@ export const DEFAULT_MESSAGE_ACTIONS = [
   }
 ];
 
-export function getComposerApplications() {
-  if(eXo && eXo.chat && eXo.chat.room && eXo.chat.room.extraApplications) {
-    return DEFAULT_COMPOSER_APPS.concat(eXo.chat.room.extraApplications);
-  } else {
-    return DEFAULT_COMPOSER_APPS;
+function getExtenstionByType(type) {
```
[the remark this answers is not in the document]
Comment (thomasdelhomenie) [reply application/src/main/webapp/vue-app/extension.js]: Yes, that's why I suggested to change to plural.

File: application/src/main/webapp/vue-app/extension.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomasdelhomenie) [reply application/src/main/webapp/vue-app/extension.js]: You mean our API can cause performance issues ?

File: application/src/main/webapp/vue-app/main.js [hunk 3]
```diff
@@ -32,17 +31,27 @@ Vue.directive('hold-tap', function (el, binding, vnode) {
...
+    });
+  }
+}
+
 exoi18n.loadLanguageAsync(lang, url).then(i18n => {
   if ($('#chatApplication').length) {
     new Vue({
       el: '#chatApplication',
-      render: h => h(ExoChatApp),
+      template: '<exo-chat-app></exo-chat-app>',
```
[the remark this answers is not in the document]
Comment (thomasdelhomenie) [reply application/src/main/webapp/vue-app/main.js]: ok. We should document somewhere the requirements to make a component overridable. WDYT ?

[the remark this answers is not in the document]
Comment (alaameddeb) [reply application/src/main/webapp/vue-app/main.js]: exist on the study, maybe should be added to the documentation

File: application/src/main/webapp/js/lib/extensionRegistry.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
+/**
```
[the remark this answers is not in the document]
Comment (thomasdelhomenie) [reply application/src/main/webapp/js/lib/extensionRegistry.js]: ok, let's try that :+1:

File: application/src/main/webapp/vue-app/main.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (thomasdelhomenie) [reply application/src/main/webapp/vue-app/main.js]: Yes, it should be in a documentation somewhere IMO. It should probably be part of the contributor guide (developer.exoplatform.org).

File: application/src/main/webapp/vue-app/extension.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (alaameddeb) [reply application/src/main/webapp/vue-app/extension.js]: using extensionRegistry.loadExtensions in the component is not performant

File: application/src/main/webapp/js/lib/extensionRegistry.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  }
+  
+  var extensionRegistry = {
+    "registerComponent": registerComponent, 
+    "loadComponents": loadComponents,
+    "registerExtension": registerExtension, 
+    "loadExtensions": loadExtensions
+  };
+  
+  // exoprt for nodejs
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/lib/extensionRegistry.js]:
```suggestion
  // export for nodejs (for tests)
```

**Pull request merged**